crates/astria-sequencer/src/mempool.rs

```diff
@@ -43,212 +55,236 @@ impl PartialOrd for TransactionPriority {
+    /// Updates the priority of the txs in the mempool based on the current state, and removes any
+    /// that are now invalid.
+    ///
+    /// *NOTE*: this function locks the mempool until every tx has been checked. This could
+    /// potentially stall consensus from moving to the next round if the mempool is large.
+    pub(crate) async fn update_priorities<F, O>(
+        &self,
+        current_account_nonce_getter: F,
+    ) -> anyhow::Result<()>
+    where
+        F: Fn(Address) -> O,
+        O: Future<Output = anyhow::Result<u32>>,
+    {
+        let mut txs_to_remove = Vec::new();
+        let mut current_account_nonces = HashMap::new();
+
+        let mut queue = self.inner.write().await;
+        for (enqueued_tx, priority) in queue.iter_mut() {
+            let address = enqueued_tx.address();
+            // Try to get the current account nonce from the ones already retrieved.
+            let current_account_nonce = if let Some(nonce) = current_account_nonces.get(&address) {
+                *nonce
+            } else {
+                // Fall back to getting via the getter and adding it to the local temp collection.
+                let nonce = current_account_nonce_getter(enqueued_tx.address())
+                    .await
+                    .context("failed to fetch account nonce")?;
+                current_account_nonces.insert(address, nonce);
+                nonce
+            };
+            match enqueued_tx.priority(current_account_nonce) {
+                Ok(new_priority) => *priority = new_priority,
+                Err(e) => {
+                    debug!(
+                        transaction_hash = %telemetry::display::base64(&enqueued_tx.tx_hash),
+                        error = AsRef::<dyn std::error::Error>::as_ref(&e),
+                         "account nonce is now greater than tx nonce; dropping tx from mempool",
+                    );
+                    txs_to_remove.push(enqueued_tx.clone());
+                }
+            };
         }
-    }
 
-    /// returns the inner mempool, write-locked.
-    /// required so that `BasicMempool::iter_mut()` can be called.
-    pub(crate) async fn write(&self) -> tokio::sync::RwLockWriteGuard<'_, BasicMempool> {
-        self.inner.write().await
+        for enqueued_tx in txs_to_remove {
+            queue.remove(&enqueued_tx);
+        }
+
+        Ok(())
     }
 
     /// returns the pending nonce for the given address,
     /// if it exists in the mempool.
-    #[instrument(skip(self), fields(address = %address))]
     pub(crate) async fn pending_nonce(&self, address: &Address) -> Option<u32> {
         let inner = self.inner.read().await;
         let mut nonce = None;
-        for (tx, priority) in inner.iter() {
-            let sender = Address::from_verification_key(tx.verification_key());
+        for (tx, _priority) in inner.iter() {
+            let sender = Address::from_verification_key(tx.signed_tx.verification_key());
```

Comment:
use `tx.address()` here

Reply:
I'll sort that while dealing with the merge conflicts, as `Address::from_verification_key` is now gone from `main` :)